### ISSUE TITLE
mypy recursively follows relative symlinks to the package root

### ISSUE THREAD

**Issue by droserasprout** [python/mypy#15863]
**Bug Report**

I use a weird hack to make package importable when workdir is it's root:

```shell
python -c 'import my_package'
cd my_package
ln -s . my_package
python -c 'import my_package'
```

I will probably burn in hell for this hack, but it works surprisingly well, both in runtime and when using linters. However, mypy dives into this symlink over and over checking the same files until the recursion limit. The only downside is execution time, but on large projects it can grow significantly.

**To Reproduce**

```shell
# Create a package `mypy_test`
$ mkdir mypy_test
$ touch mypy_test/__init__.py

# Check it
$ python -c 'import mypy_test'
$ mypy mypy_test
Success: no issues found in 1 source file

# Create a relative symlink to the project root
$ cd mypy_test
$ ln -s . mypy_test
$ file mypy_test 
mypy_test: symbolic link to .

# Check the package again
$ python -c 'import mypy_test'
$ mypy .
Success: no issues found in 41 source files
```

**Expected Behavior**

mypy detects the loop and doesn't enter it.

**Actual Behavior**

mypy follows the symlink checking the same files until hitting recursion limit (I guess). Here's how to confirm it:

```shell
$ mypy -v . |& grep 'Metadata fresh for mypy_test.' | head -n 5 
LOG:  Metadata fresh for mypy_test: file ./__init__.py
LOG:  Metadata fresh for mypy_test.mypy_test: file ./mypy_test/__init__.py
LOG:  Metadata fresh for mypy_test.mypy_test.mypy_test: file ./mypy_test/mypy_test/__init__.py
LOG:  Metadata fresh for mypy_test.mypy_test.mypy_test.mypy_test: file ./mypy_test/mypy_test/mypy_test/__init__.py
LOG:  Metadata fresh for mypy_test.mypy_test.mypy_test.mypy_test.mypy_test: file ./mypy_test/mypy_test/mypy_test/mypy_test/__init__.py
```

**Your Environment**

- Mypy version used: mypy 1.5.0 (compiled: yes)
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: Python 3.11.4 (main, Jun  7 2023, 17:17:32) [GCC 13.1.1 20230525] on linux
